### ISSUE TITLE
Unable to use signtool /ds option with the CNG provider

### ISSUE THREAD

**Issue by obones** [GoogleCloudPlatform/kms-integrations#20]
Hello,

Because of the absence of an x86 version of the provider (See #18), I'm trying a workaround by using the following steps:

1. get the file digest by using the `/dg` option

        signtool sign /q /f "%CertificatePublicKeyFilename%" /du %SignatureUrl%  /fd sha256 /dg . "%FileToSign%"

2. Sign the generated digest with the CNG provider and the `/ds` option

        signtool sign /q /f "%CertificatePublicKeyFilename%"  /csp "%GoogleCSPName%" /kc "%GoogleCloudKeyVersionFullPath%" /ds "%DigestFile%"

3. Integrate the signed digest with the `/di` option

        signtool sign /q /f "%CertificatePublicKeyFilename%" /di . "%FileToSign%"

4. Timestamp the generated signature

        signtool timestamp /q /tr %SignatureTimeStampURL% /td sha256 "%FileToSign%"

When using a certificate whose private key is stored on a physical usb token (Yubikey), this works fine.
But when I use the CNG provider, I get the following error:

    I0927 12:56:16.817818   31368 logging.cc:81] returning 0x80090027 from SignHashFn due to status INVALID_ARGUMENT: at bridge.cc:421: unsupported pPaddingInfo [type.googleapis.com/kmscng.StatusDetails='SECURITY_STATUS=0x80090027']

Looking at [bridge.cc:421](https://github.com/GoogleCloudPlatform/kms-integrations/blob/master/kmscng/main/bridge.cc#L421) the error seems to be created on purpose but I'm not sure what to make of this.
I would simply remove the test as it seems `pPaddingInfo` is not used at all in that function, but this must be a very naive approach.

Any help would be most welcome.

**Comment by tdbhacks** [comment]
Hello,

I am not familiar with this flow, but per the [CNG spec](https://learn.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptsignhash#parameters), the `pPaddingInfo` parameter should really only be used when passing in structs of type [BCRYPT_PKCS1_PADDING_INFO](https://learn.microsoft.com/en-us/windows/win32/api/bcrypt/ns-bcrypt-bcrypt_pkcs1_padding_info) or [BCRYPT_PSS_PADDING_INFO](https://learn.microsoft.com/en-us/windows/win32/api/bcrypt/ns-bcrypt-bcrypt_pss_padding_info).

I guess the parameter description leaves some room for interpretation because it says `"This parameter is only used with asymmetric keys and must be NULL otherwise"`, and ECDSA is an asymmetric algorithm, but we are expecting it to be NULL in our provider because it doesn't really make sense to specify it for ECDSA algorithms (unless my reading of the spec is wrong).

Just to confirm, are you trying to use this with an ECDSA key or an RSA-PKCS1 / RSA-PSS key? See the [limitations section](https://github.com/GoogleCloudPlatform/kms-integrations/blob/master/kmscng/docs/user_guide.md#limitations). If you are indeed using an ECDSA key, I'll tweak the provider logging when I have some time and try to go through the same flow to understand what it's trying to set it to. Thanks for bringing this up!

**Comment by obones** [comment]
To me this is an ECDSA key:
![image](https://github.com/GoogleCloudPlatform/kms-integrations/assets/1314739/cf5dcfb9-1989-477b-9b34-c9304d2cdade)

And the CNG provider works if I don't use the `/ds` option but rather have `signtool` do all actions at once.

**Comment by tdbhacks** [comment]
Indeed, I found some time to reproduce this with my own EC-P256 key and I also ran into the same error.

Played with it for a bit and tweaked our logging to check what is being passed to the provider: it looks like SignTool is trying to use the `BCRYPT_PAD_PKCS1` flag in combination with a [BCRYPT_PKCS1_PADDING_INFO](https://learn.microsoft.com/en-us/windows/win32/api/bcrypt/ns-bcrypt-bcrypt_pkcs1_padding_info) struct with `pszAlgId` == `SHA256`, so at this point I'm pretty confused because this whole flow should only apply to RSA-PKCS1 keys...

We should investigate further.

**Comment by ysichrisdag** [comment]
Is there any plan to address this bug?  Really need either this one addressed or https://github.com/GoogleCloudPlatform/kms-integrations/issues/18

**Comment by tdbhacks** [comment]
@obones @ysichrisdag as I was going through the open issues of the repo, I figured I'd try out this flow again now that RSA4096-PKCS1 support has been merged (#29). I haven't tried ECDSA, I assume the bug is still present and I don't think we can do much about that, but I believe I was able to sign something successfully using `RSA_SIGN_PKCS1_4096_SHA256`!

Is ECDSA a requirement? If not, it would be great if you could test this out with RSA

**Comment by obones** [comment]
> I haven't tried ECDSA, I assume the bug is still present and I don't think we can do much about that
Well, this issue is here just because I tried a workaround issue #18 but if that other issue was to be addressed, then this one could go way down in the priority list.

> Is ECDSA a requirement?
Yes, it is because it comes from my signing key provider as I require a signing key backed by the PKI recognized by the Windows operating system.

**Comment by obones** [comment]
FWIW, I tried with signtool from SDK 10.0.26100.0 but this did not change the situation

**Comment by obones** [comment]
I just updated the CNG provider to the latest version, and the error message is gone.

Sadly, the exe file is not properly signed, the file property dialog shows this error:

<img width="406" height="484" alt="Image" src="https://github.com/user-attachments/assets/10607405-fd6f-443c-b880-81cbcb1b992b" />

There are traces of a similar issue at [StackOverflow](https://stackoverflow.com/questions/76207339/authenticode-sign-the-hash-of-a-file-with-an-external-tool) with the following comment:

> The solution is to add the SHA header to the digest before signing it.

But prepending the SHA512 to the `.dig` is not an option, this gives a new error:

```
logging.cc:81] returning 0x80090027 from SignHashFn due to status INVALID_ARGUMENT: at sign_utils.cc:316: provided digest has incorrect size (got 80, want 32) [type.googleapis.com/kmscng.StatusDetails='SECURITY_STATUS=0x80090027']
```

So clearly, this was not what is meant by the comment, but I can't figure out what to do.
Note: I know this is not really related to CNG Provider but I'm posting here just in case someone knows what to do to make a valid signature when using the `/dg`, `/ds` and `/di` options in sequence.
